### ISSUE TITLE
Proposal: IPFS in web browser (HTTP compatible)

### ISSUE THREAD

**Issue by kehao95** [ipfs/ipfs#291]
Hi, I've been discuses about the IPFS in native browser support in the this thread 
[web-browser-with-integrated-ipfs-node-support-for-browser-cache - discuss.ipfs.io](https://discuss.ipfs.io/t/web-browser-with-integrated-ipfs-node-support-for-browser-cache/1799) where @jamaalsawyerd put up an ideas, that the browser can serve

We would love to see IPFS take the place of HTTP in certain areas. Where in the future the browsers are supporting ipfs and every user while surfing the internet, can contribe and retrive content in ipfs. And I've come up some ideas I'd like to throw some light on here.

So how can we use ipfs while being compatible with HTTP. How can the client know the multi-hash 
 I've some ideas may work in limit situations.

## [1]. Expose multi-hash in files’ url parameters
This may be the simplest one that the sever can prove the file’s ipfs hash in the url parms. This can be done in many links in HTML, whether it’s image source, css source or script source, etc. So the client can scan the url and decide to get the file from ipfs network or preform a direct http request.

## [2]. Using HTTP 300:
There is potential compatibility in HTTP Stander in HTTP 3xx (Redirection). The Redirection is not limit to HTTP but can be any URI (ipfs included) Especially for HTTP 300
> The HTTP 300 Multiple Choices redirect status response code indicates that the request has more than one possible responses. The user-agent or the user should choose one of them. As there is no standardized way of choosing one of the responses, this response code is very rarely used.
When receiving an HTTP GET request for a file, instead of directly return the data, the server can choose response an HTTP 300 with 2 URIs including one direct HTTP GET link and one IPFS link. So that the client can choose IPFS to get the files in IPFS network. 
Risk: since there is no standardized way of choosing one of the responses. No sure if normal browsers without ipfs support can handle this properly.

## [3]. HTTP HEAD before GET:
Before GET a resource, the browser can send a HEAD request to get meta info about the file. The return header always includes the file's Content-Length Content-Type etc, in rare cases, the file's checksum is included as well. For IPFS support, the sever can return the IPFS hash as well. So that the browser can use the IPFS hash to get the file instead of performing a GET request.

**Comment by jamaalsawyerd** [comment]
Yes I agree with the above!

**Comment by lgierth** [comment]
Thanks for posting your ideas. Work on ipfs in web browsers is tracked in the [in-web-browsers repo](https://github.com/ipfs/in-web-browsers).

**Comment by lidel** [comment]
@kehao95  [in-web-browsers](https://github.com/ipfs/in-web-browsers/issues) is a better place for this, but to save you and others some time:

- **[1] is already done**, with a small twist: if you have permission to can add a parameter to URL, you could just host your files under `/ipfs/$cid`, right? 
And HTTP paths like that are already detected and redirected to local IPFS gateway by our browser extension before actual request is sent :) 
    See [my last comment](https://github.com/ipfs-shipyard/ipfs-companion/issues/16#issuecomment-336641844) at https://github.com/ipfs-shipyard/ipfs-companion/issues/16 – we use [is-ipfs](https://github.com/ipfs/is-ipfs) lib for this.
- **[2] probably won't happen** because we can't guarantee uniform distribution of browser requests: 
   > There is no standardized way to choose one of the responses.
   > – https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html    

  There are more reliable ways to produce similar result, eg. (i) your HTTP server can return 302 redirect to a random public gateway in round-robin fashion or (ii) you could set up [anycast at DNS level](https://en.wikipedia.org/wiki/Anycast) etc.

- **[3] most likely won't happen**  because no browser vendor will implement support for replacing `GET` with `HEAD` + `GET` – it introduces overhead via wasteful requests. I am sure a proposal to always send [currently conditionally triggered preflight  request](https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request) (even only once per every hostname) would be rejected as well. Especially when client / browser extension can perform opportunistic IPFS detection at no cost before a request is sent, as I noted in response for [1].

If you think this does not answer your questions sufficiently, please create a new issue in [in-web-browsers](https://github.com/ipfs/in-web-browsers/issues) repo or (preferably) ask them in original [discuss.ipfs.io thread](https://discuss.ipfs.io/t/web-browser-with-integrated-ipfs-node-support-for-browser-cache/1799) and `@mention` me :)

**Comment by kehao95** [comment]
@lidel Thanks for your reply. Now I'm convinced by you. But I think there is still a chance for HTTP 300. We may need some tests for browsers in the future.
